Add Voice instructions feature

Guardiola31337 opened the pull request:
### Description

Adds Voice instructions feature.
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/3639 
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/3396
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/1749
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/3835
Fixes https://github.com/mapbox/mapbox-navigation-android/issues/2025

Opening this as `Draft` to get some feedback while finishing up the missing pieces.

### Changelog

```
<changelog>Added Voice instructions feature.</changelog>
```

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt:
```diff
@@ -0,0 +1,112 @@
+                        val voiceAction = VoiceAction.ProcessVoiceResponse(response)
+                        when (val res = VoiceProcessor.process(voiceAction)) {
+                            is VoiceResult.Voice.Success -> {
+                                ThreadController.getMainScopeAndRootJob().scope.launch {
```
idea: assign each VoiceInstructions a Job, cancel would cancel the job

Guardiola31337 replied:
It's not that easy as that'd only cover the case for cancelling storing the file into disk but neither the rest of the callbacks nor the `mapboxSpeech.enqueueCall`.

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt:
```diff
@@ -0,0 +1,112 @@
+     * Given the [Announcement] the method may cleanup any associated files previously generated.
+     * @param announcement
+     */
+    override fun clean(announcement: Announcement) {
```
Should this be `VoiceInstructions` or is there a way for the caller to get the `Announcement`?

Guardiola31337 replied:
https://github.com/mapbox/mapbox-navigation-android/blob/9d9590ccafcb89459af455a9fe70b84435ce81d4/test-app/src/main/java/com/mapbox/navigation/examples/core/VoiceActivity.kt#L137-L141

kmadsen replied:
should `clean` be `internal` then?

Guardiola31337 replied:
Nah, it's up to the client to clean up the resources. Was thinking about doing this automatically but then the `SpeechPlayer` needs to know about the `SpeechApi` and wanted to avoid that coupling.

kmadsen replied:
> Nah, it's up to the client to clean up the resources. 

I'm not seeing why we want to make require the client to do this yet 🤔

LukasPaczos replied:
`SpeechPlayer` would only need to know about the `MapboxSpeechFileProvider`. Maybe that's a worthy tradeoff? I agree with @kmadsen that since we don't have caching mechanism, we should be cleaning up after ourselves automatically.

LukasPaczos replied:
But I also see that if someone would like to use the `SpeechPlayer` without our `SpeechApi`, then we would be removing their files which isn't ideal. The way the implementation is decoupled does make sense that users need to clean up manually, it just isn't convenient. We have an argument both ways I guess.

Guardiola31337 replied:
Correct, I wanted to avoid that coupling.

Guardiola31337 replied:
Also, we were discussing that it may be a use case for the clients that would like to keep the files ¯\\\_(ツ)\_/¯

LukasPaczos replied:
I don't think reusability is a valid scenario right now since we assign random IDs which reset with every instantiation of the SDK.

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt:
```diff
@@ -0,0 +1,112 @@
+            val cacheDirectory = context.applicationContext.cacheDir
+            val instructionsCacheDirectory = File(cacheDirectory, MAPBOX_INSTRUCTIONS_CACHE)
+                .also { it.mkdirs() }
+            File(instructionsCacheDirectory, "${retrieveUniqueId()}$MP3_EXTENSION")
```
Is this file cleaned up anywhere? How large can it get?

kmadsen replied:
nevermind, after it is played it is deleted

Guardiola31337 replied:
Refs. https://github.com/mapbox/mapbox-navigation-android/pull/3977/files#r574583524

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,129 @@
+ * @property context Context
+ * @property language String
+ */
+internal class MapboxOnboardSpeechPlayer(
```
Considering that this is an internal class, and it is only used by the MapboxHybridSpeechPlayer

Consider not making it implement the SpeechPayer. That way the interactions are specific for MapboxHybridSpeechPlayer.

For example, `MapboxOffboardSpeechPlayer` can only play a File. So it fails when you send it a `SpeechState.Play` without a `File`.

Think it can be more explicit: 
-- MapboxHybridSpeechPlayer.play(SpeechState.Play)
---- MapboxOffboardSpeechPlayer.play(file: File)

Guardiola31337 replied:
It doesn't fail. Basically, it doesn't play 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/9d9590ccafcb89459af455a9fe70b84435ce81d4/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOffboardSpeechPlayer.kt#L43-L46

Actually, we're checking if the file exists or not in `MapboxHybridSpeechPlayer` to dispatch the playing to `MapboxOffboardSpeechPlayer` or `MapboxOnboardSpeechPlayer` https://github.com/mapbox/mapbox-navigation-android/blob/9d9590ccafcb89459af455a9fe70b84435ce81d4/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxHybridSpeechPlayer.kt#L86-L88

kmadsen replied:
yeah this is what I meant by fail, is that it doesn't play. but looks like it is an internal class now so marking it resolved

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOffboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,120 @@
+    private var mediaPlayer: MediaPlayer? = null
+    private var volumeLevel: Float = DEFAULT_VOLUME_LEVEL
+    private lateinit var clientCallback: SpeechPlayerCallback
+    private var currentPlay: SpeechState.Play = SpeechState.Play(Announcement("", null, null))
```
```suggestion
    private var currentPlay: SpeechState.Play? = null
```

up to you but i prefer `null`, that way the compiler will help make sure empty announcements are properly handled

Guardiola31337 replied:
The idea here was that `currentPlay` couldn't be `null` 😅

kmadsen replied:
what happens when it's empty? do we need to filter them? does the caller need to filter them?

Guardiola31337 replied:
https://github.com/mapbox/mapbox-navigation-android/blob/9d9590ccafcb89459af455a9fe70b84435ce81d4/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOffboardSpeechPlayer.kt#L43-L44

LukasPaczos replied:
> that way the compiler will help make sure empty announcements are properly handled

I like that as well :+1:

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxHybridSpeechPlayer.kt:
```diff
@@ -0,0 +1,91 @@
+        clientCallback = callback
+        queue.add(state)
+        if (queue.size == 1) {
+            play()
```
if there are threading issues here, e.g., play needs to be called from a background thread

the play could only be triggered by the queue, and calling play only adds to the play queue

kmadsen replied:
`@UiThread`

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,129 @@
+
+    private companion object {
+        private const val DEFAULT_UTTERANCE_ID = "default_id"
+        private const val DEFAULT_VOLUME_LEVEL = 0.5f
```
nit: Should these be `SpeechPlayerOptions`?

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/VoiceProcessor.kt:
```diff
@@ -0,0 +1,47 @@
+
+    private fun prepareRequest(instruction: VoiceInstructions): VoiceResult.VoiceRequest {
+        val request = MapboxSpeech.builder()
+            .instruction(instruction.ssmlAnnouncement() ?: instruction.announcement() ?: "")
```
should the last branch (`?: ""`) be an error?

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxHybridSpeechPlayer.kt:
```diff
@@ -0,0 +1,91 @@
+ * [MapboxOffboardSpeechPlayer] speech players.
+ * @property context Context
+ * @property accessToken String
+ * @property language String
```
What's the schema?

Guardiola31337 replied:
What do you mean by schema here? 🤔

LukasPaczos replied:
How should the passed `string`  look like, is it a specific iso code?

Guardiola31337 replied:
Ohh for the language. Got it. It needs to be a `Locale` language (ISO 639). Will add a comment, good point 👍

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxHybridSpeechPlayer.kt:
```diff
@@ -0,0 +1,91 @@
+        MapboxOffboardSpeechPlayer(context, accessToken, language)
+    private val onboardSpeechPlayer: MapboxOnboardSpeechPlayer =
+        MapboxOnboardSpeechPlayer(context, language)
+    private var clientCallback: SpeechPlayerCallback? = null
```
`clientCallback` should be paired with the internal `play` request, we cannot store it as a single field, it's going to be overwritten by queued announcments.

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOffboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,120 @@
+     * and optionally a synthesized speech mp3.
+     * @param callback
+     */
+    override fun play(state: SpeechState.Play, callback: SpeechPlayerCallback) {
+        clientCallback = callback
+        currentPlay = state
```
How about we assert here that `play` is never invoked when the previous announcement has not finished playing? This way we'll find bugs much quicker.

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,129 @@
+     * and optionally a synthesized speech mp3.
+     * @param callback
+     */
+    override fun play(state: SpeechState.Play, callback: SpeechPlayerCallback) {
+        clientCallback = callback
+        currentPlay = state
```
Same as above.

LukasPaczos commented:
This doesn't currently resolve https://github.com/mapbox/mapbox-navigation-android/issues/822 (as mentioned in the OP).

Guardiola31337 commented:
> This doesn't currently resolve https://github.com/mapbox/mapbox-navigation-android/issues/822 (as mentioned in the OP).

Was going to close that one altogether as that doesn't make sense. The recommended way by the docs is to recreate the `MediaPlayer` so that ticket doesn't make sense. I added it here so that was closed automatically. But will add a comment to the ticket and close it manually 😅

Guardiola31337 commented:
> Was going to close that one altogether as that doesn't make sense. The recommended way by the docs is to recreate the `MediaPlayer` so that ticket doesn't make sense. I added it here so that was closed automatically. But will add a comment to the ticket and close it manually 😅  

Closed manually and removed it from OP 🚀

korshaknn commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,130 @@
+) : SpeechPlayer {
+
+    private var isLanguageSupported: Boolean = false
+    private val textToSpeech: TextToSpeech = TextToSpeech(context.applicationContext) { status ->
+        val ableToInitialize = status == TextToSpeech.SUCCESS
+        if (!ableToInitialize) {
+            return@TextToSpeech
+        }
+        initializeWithLanguage(Locale(language))
+    }
```
```kotlin
private val textToSpeech = TextToSpeech(context.applicationContext) { status ->
        if (status == TextToSpeech.SUCCESS) {
            initializeWithLanguage(Locale(language))
        }
}
```

korshaknn commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    private fun initializeWithLanguage(language: Locale) {
+        val isLanguageAvailable =
+            textToSpeech.isLanguageAvailable(language) == TextToSpeech.LANG_AVAILABLE
+        if (!isLanguageAvailable) {
+            return
+        }
+        isLanguageSupported = true
```
```kotlin
isLanguageSupported =
            textToSpeech.isLanguageAvailable(language) == TextToSpeech.LANG_AVAILABLE
if (!isLanguageSupported) {
   return
}
```

korshaknn commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,130 @@
+        initializeWithLanguage(Locale(language))
+    }
+    private var volumeLevel: Float = DEFAULT_VOLUME_LEVEL
+    private lateinit var clientCallback: SpeechPlayerCallback
```
```suggestion
    private var clientCallback: SpeechPlayerCallback? = null
```

korshaknn commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    private fun play(announcement: String) {
+        val bundle = Bundle()
+        bundle.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, volumeLevel)
```
val bundle = Bundle().aplly{putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, volumeLevel)}

Guardiola31337 commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/VoiceProcessor.kt:
```diff
@@ -0,0 +1,47 @@
+
+    private fun prepareRequest(instruction: VoiceInstructions): VoiceResult.VoiceRequest {
+        val request = MapboxSpeech.builder()
+            .instruction(instruction.ssmlAnnouncement() ?: instruction.announcement() ?: "")
```
I believe you meant

```kotlin
val request = MapboxSpeech.builder()
     .instruction(announcement)
     .textType(type)
```

Guardiola31337 commented:
@kmadsen @LukasPaczos @korshaknn addressed feedback, this is ready for another round of 👀

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxOnboardSpeechPlayer.kt:
```diff
@@ -0,0 +1,134 @@
+    override fun play(state: SpeechState.ReadyToPlay, callback: SpeechPlayerCallback) {
+        clientCallback = callback
+        check(currentPlay == null) {
+            "Only one announcement can be played at a time."
```
The queue prevents this from happening in the hybrid player, right?

Guardiola31337 replied:
Correct. That's addressing @LukasPaczos's comment https://github.com/mapbox/mapbox-navigation-android/pull/3977#discussion_r574656472 

Since`Onboard` and `Offboard` are internal implementations, we can consider that if `play` is invoked while another instruction is playing a bug, as it's controlled by us in the `Hybrid`.

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxHybridSpeechPlayer.kt:
```diff
@@ -0,0 +1,100 @@
+    private var localCallback: SpeechPlayerCallback = object : SpeechPlayerCallback {
+        override fun onDone(state: SpeechState.DonePlaying) {
+            val announcement = queue.poll()
+            val clientCallback = clientCallbacks.first
```
Are we only calling onDone for the first callback? Seems like it should remove from `clientCallbacks` 🤔

Guardiola31337 replied:
Great catch! While working with the queue forgot that `first` doesn't remove from the list 😅

Guardiola31337 replied:
Will replace by `removeFirst` 👍

Guardiola31337 replied:
Actually, I'm going to replace it by `ConcurrentLinkedQueue` (and will use `poll`) to avoid any potential races.

kmadsen commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxHybridSpeechPlayer.kt:
```diff
@@ -0,0 +1,100 @@
+        MapboxOffboardSpeechPlayer(context, accessToken, language)
+    private val onboardSpeechPlayer: MapboxOnboardSpeechPlayer =
+        MapboxOnboardSpeechPlayer(context, language)
+    private val clientCallbacks: LinkedList<SpeechPlayerCallback> = LinkedList()
```
Why is this a LinkedList? It's this supposed to be in sync with the `Queue<SpeechState.ReadyToPlay>`?

Guardiola31337 replied:
We know for sure that announcements are played in order so an ordered list in which the elements are FIFO should be enough. Does that make sense?

Guardiola31337 replied:
https://github.com/mapbox/mapbox-navigation-android/pull/3977#discussion_r576439038

kmadsen commented on circle.yml:
```diff
@@ -236,6 +236,8 @@ commands:
       - store-results:
           module_target: "libnavui-tripprogress"
+      - store-results:
+          module_target: "libnavui-voice"
```
why do we want this to be another module? in other words, what exactly makes voice a good candidate for modularity?

is this to help compile time, will customers create their own voice modules, does the sdk support "bring your own voice instructions", or something else?

Guardiola31337 commented:
Still working on tests but this is definitely ready for review. Removing `Draft` here.

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceInstructionsPlayer.kt:
```diff
@@ -0,0 +1,102 @@
+) : VoiceInstructionsPlayer {
+
+    private val queue: Queue<SpeechState.ReadyToPlay> = ConcurrentLinkedQueue()
+    // for testing
+    internal var filePlayer: VoiceInstructionsFilePlayer =
+        VoiceInstructionsFilePlayer(context, accessToken, language)
+    // for testing
+    internal var textPlayer: VoiceInstructionsTextPlayer =
+        VoiceInstructionsTextPlayer(context, language)
```
Same as above with the internal `object` provider instead of `internal var` fields.

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceInstructionsPlayer.kt:
```diff
@@ -0,0 +1,102 @@
+    private var localCallback: VoiceInstructionsPlayerCallback =
+        object : VoiceInstructionsPlayerCallback {
+            override fun onDone(state: SpeechState.DonePlaying) {
+                val announcement = queue.poll()
+                val clientCallback = clientCallbacks.poll()
```
Wondering if it wouldn't be easier to have a data object that contains both `announcement` and `clientCallback` and maintain only one queue.

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsFilePlayer.kt:
```diff
@@ -0,0 +1,129 @@
+        val file = state.announcement.file
+        if (file != null && file.canRead()) {
+            play(file)
+        }
```
```suggestion
        } else {
            Log.e()
            donePlaying()
        }
```

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsFilePlayer.kt:
```diff
@@ -0,0 +1,129 @@
+    private fun addListeners() {
+        mediaPlayer?.run {
+            setOnErrorListener { _, _, _ ->
+                false
```
```suggestion
                Log.e()
                onDonePlaying()
```

Guardiola31337 replied:
`donePlaying` is not necessary, returning `false` will cause the `OnCompletionListener` to be called.

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsTextPlayer.kt:
```diff
@@ -0,0 +1,134 @@
+        val announcement = state.announcement.announcement
+        if (isLanguageSupported && announcement.isNotBlank()) {
+            play(announcement)
+        }
```
```suggestion
        } else {
            donePlaying()
        }
```

LukasPaczos commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsTextPlayer.kt:
```diff
@@ -0,0 +1,134 @@
+            textToSpeech.isLanguageAvailable(language) == TextToSpeech.LANG_AVAILABLE
+        if (!isLanguageSupported) {
+            return
+        }
```
```suggestion
        } else {
            donePlaying()
        }
```

LukasPaczos replied:
Missing `donePlaying` call.

LukasPaczos replied:
This is init time :+1:

Guardiola31337 commented on libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechApi.kt:
```diff
@@ -0,0 +1,98 @@
+                callback.onAvailable(
+                    SpeechState.Speech.Available(
+                        Announcement(
+                            // Can't be null as it's checked in VoiceProcessor#prepareRequest
```
`VoiceInstructions` (`mapbox-java`) ¯\\\_(ツ)\_/¯

https://github.com/mapbox/mapbox-java/blob/f16813f3489269011ebacfce8a56fa6b5b3279f4/services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/VoiceInstructions.java#L42-L60

Good opportunity to ask @danpat and @mapbox/navnative if these can be `null`. Can they?

LukasPaczos reviewed (approved):
Great work here!

Let's merge to start internal integration and follow up on:
- https://github.com/mapbox/mapbox-navigation-android/pull/3977/files#r578458964 queue refactor.
- Any missing unit tests.
- A high-level instrumentation test.
